cmd/ferretdb/ping.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// ping creates connection to FerretDB instance specified by the flags, and runs `ping` command against it.
+// The check is only executed if --setup-database flag is set.
+func ping() {
+	logger := setupLogger(cli.Log.Format, "")
+	checkFlags(logger)
+
+	l := logger.Sugar()
+
+	if cli.Setup.Database == "" {
+		l.Info("Setup database not specified - skipping ping.")
+		return
+	}
+
+	var urls []string
+
+	if cli.Listen.Addr != "" {
+		host, port, err := net.SplitHostPort(cli.Listen.Addr)
+		if err != nil {
+			l.Fatal(err)
+		}
+
+		l.Debugf("--listen-addr flag is set. Ping to %s will be performed.", cli.Listen.Addr)
+
+		if host == "" {
+			host = "127.0.0.1"
+
+			l.Debugf("Host not specified, defaulting to %s.", host)
+		}
+
+		u := &url.URL{
+			Scheme: "mongodb",
+			Host:   fmt.Sprintf("%s:%s", host, port),
+			Path:   cli.Setup.Database,
+			User:   url.UserPassword(cli.Setup.Username, cli.Setup.Password),
+		}
+
+		urls = append(urls, u.String())
+	}
+
+	if cli.Listen.Unix != "" {
+		l.Debugf("--listen-unix flag is set. Ping to %s will be performed.", cli.Listen.Unix)
+
+		urls = append(urls, "mongodb://"+url.PathEscape(cli.Listen.Unix))
+	}
+
+	if len(urls) == 0 {
+		l.Info("Neither --listen-addr nor --listen-unix flags were specified - skipping ping.")
+		return
+	}
+
+	for _, u := range urls {
+		l.Debugf("Pinging %s...", u)
+
+		ctx, cancel := context.WithTimeout(context.Background(), cli.Setup.Timeout)
+		defer cancel()
+
+		client, err := mongo.Connect(ctx, options.Client().ApplyURI(u))
+		if err != nil {
+			l.Fatal(err)
+		}
+
+		if err = client.Ping(ctx, nil); err != nil {
+			l.Fatal(err)
+		}
+
+		if err = client.Disconnect(ctx); err != nil {
+			l.Fatal(err)
+		}
+
+		l.Infof("Ping to %s successful.", u)
```

Comment:
We should not log secrets (such as passwords) on level higher than debug

Reply:
I think there is one too many log I think it was commented before, at the moment see for listen addr 
`l.Debugf("--listen-addr flag is set. Ping to %s will be performed.", cli.Listen.Addr)`
`l.Debugf("Pinging %s...", u)`
`l.Infof("Ping to %s successful.", u)`

Can we have one for starting such as `Pinging blah blah...` and one for success such as `Ping success blah blah`?

Reply:
@chilagrow https://github.com/FerretDB/FerretDB/pull/4364#discussion_r1647816460
I will send a PR codifying that

Reply:
@AlekSi Oops, fixed

@chilagrow 2 of the lines you mentioned are only displayed in the debug mode. Normally (on the Info level) we only log the ping success, or when no ping is executed - the skip message.

Reply:
Shouldn't we avoid displaying password even in debug?

Reply:
No, everything is a fair game on debug level. First, all information is needed for debugging. Second, we log full requests and responses on debug level that are much more secret than anything else, as they contain actual user data.